// redux-firestore/lib/reducers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "cacheReducer", {
  enumerable: true,
  get: function get() {
    return _cacheReducer.default;
  }
});
Object.defineProperty(exports, "crossSliceReducer", {
  enumerable: true,
  get: function get() {
    return _crossSliceReducer.default;
  }
});
Object.defineProperty(exports, "dataReducer", {
  enumerable: true,
  get: function get() {
    return _dataReducer.default;
  }
});
Object.defineProperty(exports, "errorsReducer", {
  enumerable: true,
  get: function get() {
    return _errorsReducer.default;
  }
});
Object.defineProperty(exports, "listenersReducer", {
  enumerable: true,
  get: function get() {
    return _listenersReducer.default;
  }
});
Object.defineProperty(exports, "orderedReducer", {
  enumerable: true,
  get: function get() {
    return _orderedReducer.default;
  }
});
Object.defineProperty(exports, "queriesReducer", {
  enumerable: true,
  get: function get() {
    return _queriesReducer.default;
  }
});
Object.defineProperty(exports, "statusReducer", {
  enumerable: true,
  get: function get() {
    return _statusReducer.default;
  }
});
var _dataReducer = _interopRequireDefault(require("./dataReducer"));
var _errorsReducer = _interopRequireDefault(require("./errorsReducer"));
var _listenersReducer = _interopRequireDefault(require("./listenersReducer"));
var _orderedReducer = _interopRequireDefault(require("./orderedReducer"));
var _queriesReducer = _interopRequireDefault(require("./queriesReducer"));
var _statusReducer = _interopRequireDefault(require("./statusReducer"));
var _crossSliceReducer = _interopRequireDefault(require("./crossSliceReducer"));
var _cacheReducer = _interopRequireDefault(require("./cacheReducer"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
